Mathlib/RingTheory/Coalgebra/TensorProduct.lean

```diff
@@ -0,0 +1,192 @@
+/-
+Copyright (c) 2024 Amelia Livingston. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Amelia Livingston
+-/
+import Mathlib.Algebra.Category.CoalgebraCat.MonEquivalence
+
+/-!
+# Tensor products of coalgebras
+
+Given two `R`-coalgebras `M, N`, we can define a natural comultiplication map
+`Δ : M ⊗[R] N → (M ⊗[R] N) ⊗[R] (M ⊗[R] N)` and counit map `ε : M ⊗[R] N → R` induced by
+the comultiplication and counit maps of `M` and `N`. These `Δ, ε` are declared as linear maps
+in `TensorProduct.instCoalgebraStruct` in `Mathlib.RingTheory.Coalgebra.Basic`.
+
+In this file we show that `Δ, ε` satisfy the axioms of a coalgebra, and also define other data
+in the monoidal structure on `R`-coalgebras, like the tensor product of two coalgebra morphisms
+as a coalgebra morphism.
+
+## Implementation notes
+
+We keep the linear maps underlying `Δ, ε` and other definitions in this file syntactically equal
+to the corresponding definitions for tensor products of modules in the hope that this makes
+life easier. However, to fill in prop fields, we use the API in
+`Mathlib.Algebra.Category.CoalgebraCat.MonEquivalence`. That file defines the monoidal category
+structure on coalgebras induced by an equivalence with monoid objects in the opposite
+category of modules, `CoalgebraCat.instMonoidalCategoryAux`, but we do not declare this as an
+instance - we just use it to prove things. Then, we use the definitions in this file to make a
+monoidal category instance on `CoalgebraCat R` in `Mathlib.Algebra.Category.CoalgebraCat.Monoidal`
+that has simpler data.
+
+-/
+
+universe v u
+open CategoryTheory
+open scoped TensorProduct
+
+section
+
+variable {R M N P Q : Type u} [CommRing R]
```

Comment:
This is the big downside of doing everything via the categorical API; it forces universe monomorphism.

Reply:
(I note that https://github.com/ImperialCollegeLondon/FLT/blob/main/FLT/for_mathlib/Coalgebra/TensorProduct.lean contains a more direct approach without going via category theory, which preserves universe polymorphism)

Reply:
Yeah that is a shame but I'd been hoping it's not too much of an issue in practice. I like the idea of getting non-category-theory library API using category theory (but only in proofs so for the most part you can't tell)... and I thought this was worth the drawback of universe monomorphism. But I'm aware of the other formalisation and if you think it's more suited to mathlib I'll close my PRs.

Reply:
I think the category-free version is more suited, yes. I'm glad to see you removed the "I'll close my PRs" section of your comment; the categorical API is still worth having, just not for the purpose of obtaining this instance.

Reply:
Okay. yeah I guess I could just close this one and remove the second half of #11974 (or actually just replace that PR with the equivalence `CoalgebraCat R ≌ Comon_ (ModuleCat R)`). And base the others on the category-free instance when it exists.

Reply:
After talking to Amelia, I am quite keen to merge this after the merge-conflict is resolved. We will, of course, when it is ready, provide the universe generalization enabled by doing this non-categorically, but this PR is blocking important work that Amelia is doing, and I don't want to block ongoing work for generalisations that aren't quite ready to merge.

Reply:
> I am quite keen to merge this after the merge-conflict is resolved. 

note there is still a dependent PR

Reply:
There is, sorry about this - if it's indeed ok to proceed with these 2 PRs I'm going to refactor the dependent PR to use `CoalgebraCat R ≌ Comon_ (ModuleCat R)`. I'm at a workshop so didn't finish that yet, but I have a free afternoon today.

Reply:
Modulo CI I've hopefully finished refactoring this and #11974 to use comonoid objects instead of monoid objects. But fixing the second half of `Mathlib.Algebra.Category.Coalgebra.ComonEquivalence` gave me the impression the code is quite brittle. I'm sorry this situation is a bit of a mess - it could all have been avoided if I'd made a Zulip thread asking for advice about 2 months ago so I've learnt my lesson. If it ultimately doesn't get merged, or it gets merged and then gets entirely replaced, I won't be offended haha